test/utils/floating_point_utils.hpp

```diff
@@ -0,0 +1,81 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2020 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <cmath>
+#include <cstdint>
+#include <limits>
+#include <type_traits>
+
+namespace fizzy::test
+{
+/// Simple implementation of C++20's std::bit_cast.
+template <typename DstT, typename SrcT>
+DstT bit_cast(SrcT x) noexcept
+{
+    DstT z;
+    static_assert(sizeof(x) == sizeof(z));
+    __builtin_memcpy(&z, &x, sizeof(x));
+    return z;
+}
+
+
+template <typename T>
+struct FP
+{
+    static_assert(std::is_floating_point_v<T>);
+
+    /// Shortcut to numeric_limits.
+    using limits = std::numeric_limits<T>;
+
+    /// The unsigned integer type matching the size of this floating-point type.
+    using uint_type = std::conditional_t<std::is_same_v<T, float>, uint32_t, uint64_t>;
+    static_assert(sizeof(T) == sizeof(uint_type));
+
+    /// The number of mantissa bits in the binary representation.
+    static constexpr auto num_mantissa_bits = limits::digits - 1;
+
+    /// The binary mask of the mantissa part of the binary representation.
+    static constexpr auto mantissa_mask = (uint_type{1} << num_mantissa_bits) - 1;
+
+    /// The payload of the canonical NaN (only the top bit set).
+    /// See: https://webassembly.github.io/spec/core/syntax/values.html#canonical-nan.
+    static constexpr auto canon = uint_type{1} << (num_mantissa_bits - 1);
```

Comment:
Why not `nan_canonical_payload`?

Reply:
Can be, but is rather annoying in places where it is used. And wasm spec names this constants as `canon`: https://webassembly.github.io/spec/core/syntax/values.html#aux-canon

Reply:
Ok, fine to keep `canon`